flow360/component/flow360_params/unit_system.py

```diff
@@ -0,0 +1,589 @@
+
+
+class DimensionedType(ValidatedType):
+    """
+    :class: Base class for dimensioned values
+    """
+
+    dim = None
+    dim_name = None
+
+    @classmethod
+    def validate(cls, value):
+        """
+        Validator for value
+        """
+        if unit_system_manager.current:
+            if isinstance(value, numeric_type):
+                unit = unit_system_manager.current[cls.dim_name]
+                return value * unit
+        if not _has_dimensions(value, cls.dim):
+            raise TypeError(f"arg '{value}' does not match {cls.dim}")
+        return value
+
+    # pylint: disable=invalid-name
+    @classproperty
+    def NonNegative(self):
+        """
+        Shorthand for a ge=0 constrained value
+        """
+        return self.Constrained(ge=0)
+
+    # pylint: disable=invalid-name
+    @classproperty
+    def Positive(self):
+        """
+        Shorthand for a gt=0 constrained value
+        """
+        return self.Constrained(gt=0)
+
+    # pylint: disable=invalid-name
+    @classproperty
+    def NonPositive(self):
+        """
+        Shorthand for a le=0 constrained value
+        """
+        return self.Constrained(le=0)
+
+    # pylint: disable=invalid-name
+    @classproperty
+    def Negative(self):
+        """
+        Shorthand for a lt=0 constrained value
+        """
+        return self.Constrained(lt=0)
+
+    # pylint: disable=invalid-name
+    # pylint: disable=too-many-arguments
+    @classmethod
+    def Constrained(cls, gt=None, ge=None, lt=None, le=None, allow_inf_nan=False) -> Type[float]:
+        """
+        Utility method to generate a dimensioned type with constraints based on the pydantic confloat
+        """
+
+        class _Constrained:
+            """:class: Constrained
+            note that these constrains work only for values, disregards units.
+            We cannot constrain that mass > 2kg, we can only constrain that mass.value > 2
+            """
+
+            def __init__(self, **kwargs):
+                class _ConType:
+                    type_ = pd.confloat(**kwargs)
+
+                self._con_type = _ConType
+
+            def __get_validators__(self):
+                yield self.validate
+
+            def validate(self, value):
+                """additional validator for value"""
+                dimensioned_value = cls.validate(value)
+                pd.validators.number_size_validator(dimensioned_value.value, self._con_type)
+                pd.validators.float_finite_validator(dimensioned_value.value, self._con_type, None)
+                return dimensioned_value
+
+        return _Constrained(gt=gt, ge=ge, lt=lt, le=le, allow_inf_nan=allow_inf_nan)
+
+
+class LengthType(DimensionedType):
+    """:class: LengthType"""
+
+    dim = u.dimensions.length
+    dim_name = "length"
+
+
+class MassType(DimensionedType):
+    """:class: MassType"""
+
+    dim = u.dimensions.mass
+    dim_name = "mass"
+
+
+class TimeType(DimensionedType):
```

Comment:
It would be good to split the unit system into two parts: Base and Derived
Base unit system is Mass, Length, Time, Temperature
Derived unit system is Force, Velocity, Area, Pressure, etc.

Reply:
Derived system can be created directly from Base, but maybe also have an option to override Derived

Reply:
Okay, as long as the user can specify a subset of the unit system (MLTK) and infer the others from the base using those units it is fine. The separate classes was more of a convenience as MLTK is a standard in scientific texts and our end users might be more familiar with it.